src/Clay/Grid.hs

```diff
@@ -1,58 +1,271 @@
+
+data IsSpan = Span | NoSpan
+  deriving (Show, Eq)
+
+data GridLocation
+  = GridLocation_Keyword Value
+  | GridLocation_Data GridLocationData
+
+instance Val GridLocation where
+  value (GridLocation_Keyword v) = v
+  value (GridLocation_Data d) = value d
+
+instance Auto    GridLocation where auto    = GridLocation_Keyword auto
+instance Inherit GridLocation where inherit = GridLocation_Keyword inherit
+instance Initial GridLocation where initial = GridLocation_Keyword initial
+instance Unset   GridLocation where unset   = GridLocation_Keyword unset
+
+-- See under syntax:
+-- https://developer.mozilla.org/en-US/docs/Web/CSS/grid-column-start
+-- either grid index and/or named grid area is required, but span is optional
+data GridLocationData = GridLocationData IsSpan (These Integer GridArea)
+
+instance Val GridLocationData where
+  value (GridLocationData isSpan indexAndOrGridArea) =
+    if isSpan == Span
+    then value (span :: Value, indexAndOrGridArea)
+    else value indexAndOrGridArea
+
+pattern GridIndex :: Integer -> GridLocation
+pattern GridIndex n = GridLocation_Data (GridLocationData NoSpan (This n))
+
+instance Num GridLocation where
+  -- for index literals
+  fromInteger = GridIndex
+  -- for negative index literals
+  negate (GridIndex index) = GridIndex $ negate index
+  -- in general we don't support arithmetic on this type
+  negate _ = error "negate not defined for GridLocation"
+  abs = error "abs not defined for GridLocation"
+  signum = error "abs not defined for GridLocation"
+  (+) = error "addition not defined for GridLocation"
+  (*) = error "multiplication not defined for GridLocation"
+
+-------------------------------------------------------------------------------
+
+-- | Property defines the template for grid layout
+gridTemplateAreas :: GridTemplateAreas -> Css
+gridTemplateAreas = key "grid-template-areas"
+
+newtype GridTemplateAreas = GridTemplateAreas Value
+  deriving (Val, None, Inherit, Initial, Unset)
+
+instance IsList GridTemplateAreas where
+  type Item GridTemplateAreas = [GridArea]
+  toList = error "toList GridTemplateAreas is not defined"
+  fromList = GridTemplateAreas . value . fromList'
+    where
+      fromList' :: [Item GridTemplateNamedAreas] -> GridTemplateNamedAreas
+      fromList' = fromList
+
+-- have to create a newtype to override the Val instance for lists
+newtype GridTemplateNamedAreas = GridTemplateNamedAreas { unGridTemplateNamedAreas :: [[GridArea]] }
+
+instance Val GridTemplateNamedAreas where
+  value areas =
+    let
+      rows = coerce areas :: [[Text]]
+      wrapInParens text = "\"" <> text <> "\""
+      convertRow = wrapInParens . Text.intercalate " "
+    in
+      value $
+      Text.intercalate "\n" $
+      fmap convertRow $
+      rows
+
+-- | toList will throw when your grid template areas are invalid
+instance IsList GridTemplateNamedAreas where
+  type Item GridTemplateNamedAreas = [GridArea]
+  toList = unGridTemplateNamedAreas . coerce
```

Comment:
Why do you need `coerce`?

Reply:
I'm not sure why that's here. Looks like a relic.